### ISSUE TITLE
FileSystemWatcher events raise more than once

### ISSUE THREAD

**Issue by Petermarcu** [dotnet/runtime#24079]
This issue was filed here and has some more detailed discussions there. 

https://github.com/Microsoft/dotnet/issues/347

We should see if it repro's on .NET Core and then figure out if we need a fix or if there is an existing fix we can port to .NET Framework.

**Comment by gresolio** [comment]
What do you think about the solution that uses MemoryCache with the time expiration policy?
```cs
using System;
using System.IO;
using System.Runtime.Caching;

namespace FileSystemWatcherMemoryCache
{
    /// <summary>
    /// AddOrGetExisting a file event to MemoryCache, to block/swallow multiple events
    /// Actually 'handle' event inside callback for removal from cache on event expiring
    /// </summary>
    internal class SimpleBlockAndDelayExample
    {
        private readonly MemoryCache _memCache;
        private readonly CacheItemPolicy _cacheItemPolicy;
        private const int CacheTimeMilliseconds = 500;

        // Setup a FileSystemWatcher and cache item policy shared settings
        public SimpleBlockAndDelayExample(string demoFolderPath)
        {
            _memCache = MemoryCache.Default;

            var watcher = new FileSystemWatcher()
            {
                Path = demoFolderPath,
                NotifyFilter = NotifyFilters.LastWrite,
                Filter = "*.txt"
            };

            _cacheItemPolicy = new CacheItemPolicy()
            {
                RemovedCallback = OnRemovedFromCache
            };

            watcher.Changed += OnChanged;
            watcher.EnableRaisingEvents = true;

            Console.WriteLine($"Watching for writes to text files in folder: {demoFolderPath}");
        }

        // Handle cache item expiring 
        private void OnRemovedFromCache(CacheEntryRemovedArguments args)
        {
            if (args.RemovedReason != CacheEntryRemovedReason.Expired) return;

            var e = (FileSystemEventArgs)args.CacheItem.Value;

            Console.WriteLine($"Let's now respond to the event {e.ChangeType} on {e.FullPath}");
        }

        // Add file event to cache (won't add if already there so assured of only one occurance)
        private void OnChanged(object source, FileSystemEventArgs e)
        {
            _cacheItemPolicy.AbsoluteExpiration = DateTimeOffset.Now.AddMilliseconds(CacheTimeMilliseconds);
            _memCache.AddOrGetExisting(e.Name, e, _cacheItemPolicy);
        }      
    }
}
```

References: [FileSystemWatcher is a Bit Broken](https://failingfast.io/a-robust-solution-for-filesystemwatcher-firing-events-multiple-times/) and [FileSystemWatcherMemoryCache examples](https://github.com/benbhall/FileSystemWatcherMemoryCache) by Ben Hall.

**Comment by s-a** [comment]
@gresolio Looks promising. Its OK if you work with a few files, but on heavy load processing file servers all these workarounds hit hard on CPU and memory. I reported this 2017 and there is still no confirmation and no solution. 😕 Its pretty confusing why modules in other programing languages like node js just work. I personaly doubt that this problem is groundet inside Windows.

**Comment by mguinness** [comment]
This issue isn't getting a lot of love, let's upvote to see if we can get it more attention.

**Comment by gresolio** [comment]
@s-a There is always a trade-off between simplicity and performance. I dived a little into the subject, and I see that the rabbit hole is much deeper than I expected. So far, this is the best article on monitoring files on Windows that I have come across:
[Understanding ReadDirectoryChangesW](https://qualapps.blogspot.com/2010/05/understanding-readdirectorychangesw.html) by Jim Beveridge.

**Comment by KieranDevvs** [comment]
So from what I'm reading, I'm gathering that this is an issue with the underlying Win32 API and that's why this isn't getting any attention from the .NET team?

**Comment by Jozkee** [comment]
> this is an issue with the underlying Win32 API and that's why this isn't getting any attention from the .NET team?

Yes, we need to investigate if we can improve the existing FileSystemWatcher type. And if we can't, explore other alternaties like polling which was already mentioned here and on https://github.com/dotnet/runtime/issues/17111.

**Comment by CyrilCommando** [comment]
> You'll be better served using file hashes, see [Computing File Hashes in C#](http://gigi.nullneuron.net/gigilabs/computing-file-hashes-cs/) for more info.

I can not do this because each time I initialize a filestream to feed to a hash, it triggers a Last Access change event, which then triggers another hash, which triggers a change, and so on. The lack of granularity as mentioned [here](https://github.com/microsoft/dotnet/issues/347#issuecomment-320448626) is really quite bewildering. I do wonder if I'm doing something wrong, but I absolutely must check hash on file change and it triggers infinite recursion. This leaves me wondering what in God's name is going on with Windows' file system in general, because this is far from the first file-related problem I have had during my time developing and merely using the OS at all.
